### ISSUE TITLE
GraphServiceClient ignores HttpClient.BaseAddress

### ISSUE THREAD

**Issue by natelatchaw** [microsoftgraph/msgraph-sdk-dotnet#916]
https://github.com/microsoftgraph/msgraph-sdk-dotnet/blob/9d058256a6b9eba48ba20a4d672441c4632f2a52/src/Microsoft.Graph/Generated/requests/GraphServiceClient.cs#L51-L53

When initializing a GraphServiceClient instance by providing an HttpClient as a parameter, the HttpClient instance's BaseAddress property is ignored, and instead replaced with a hardcoded Global Cloud base address. This is an issue when attempting to connect to a Sovereign Cloud (US Gov, Germany, China, etc.) as generated request URLs for Graph resources now reference the wrong API endpoint (e.g., **https://graph.microsoft.com:443/v1.0/devices** vs. **https://graph.microsoft.us:443/v1.0/devices**), resulting in a 401 Unauthorized status code. Furthermore, the Graph version parameter is also ignored and replaced with a static `v1.0` string.

```
// parameters
Guid clientID = new Guid("redacted");
Guid tenantID = new Guid("redacted");
AzureCloudInstance cloudInstance = AzureCloudInstance.AzureUsGovernment;
string GraphCloudInstance = GraphClientFactory.USGOV_Cloud;
string redirectURI = "redacted";
string clientSecret = "redacted";
string version = "beta";

// build client application
ClientApplication = ConfidentialClientApplicationBuilder
                    .Create(clientID.ToString())
                    .WithRedirectUri(redirectURI)
                    .WithAuthority(cloudInstance, tenantID)
                    .WithClientSecret(clientSecret)
                    .Build();
ClientCredentialProvider clientCredentialProvider = new ClientCredentialProvider(ClientApplication as ConfidentialClientApplication);
HttpClient httpClient = GraphClientFactory.Create(clientCredentialProvider, nationalCloud: GraphCloudInstance, version: version);
GraphServiceClient = new GraphServiceClient(httpClient);

Console.WriteLine(httpClient.BaseAddress); // https://graph.microsoft.us/beta/
Console.WriteLine(GraphServiceClient.BaseUrl); // https://graph.microsoft.com/v1.0
``` 
[AB#8309](https://microsoftgraph.visualstudio.com/0985d294-5762-4bc2-a565-161ef349ca3e/_workitems/edit/8309)

**Comment by andrueastman** [comment]
@natelatchaw Thanks for raising this issue. 
The base address should definitely be pulled in from the http client when it can so as to avoid this scenario.

**Comment by andrueastman** [comment]
Closed via #959

**Comment by tjmoore** [comment]
While the fix is a workaround, it surely could have been fixed with the constructor reading the address from httpClient?

e.g.
```csharp
public GraphServiceClient(HttpClient httpClient)
   : base(httpClient.BaseAddress.ToString(), httpClient)
```

else the call to this is going to be

```csharp
GraphServiceClient = new GraphServiceClient(httpClient, httpClient.BaseAddress.ToString());
```

Alternatively can we skip HttpClient entirely and call...

```csharp
GraphServiceClient = new GraphServiceClient(clientCredentialProvider, nationalCloud: GraphCloudInstance, version: version);
```

As GraphServiceClient has a constructor for the auth provider, which is what I'm using, but to set the cloud instance I need to use GraphClientFactory.Create.

Note, I'm currently doing

```csharp
var httpClient = GraphClientFactory.Create(clientCredentialProvider, nationalCloud: graphCloudInstance);
new GraphServiceClient(httpClient.BaseAddress.ToString(), clientCredentialProvider);
```
which appears to work without this fix, but just passing in httpClient to GraphServiceClient is better, or not having to create httpClient and just pass the cloud instance and auth provider may be better still.

p.s. inconsistencies - GraphServiceClient with HttpClient has baseUrl as second argument and a default. With auth provider argument baseUrl is first argument and no default.
